torch/distributed/elastic/supervisor/launchers.py

```diff
@@ -0,0 +1,339 @@
+from torch.distributed.elastic.utils.logging import get_logger
+
+from . import as_completed, Context, Future, Host, Process, wait_on
+
+
+PORT = 55555
+REGISTRY_LAUNCHERS_METADATA_GROUP_KEY = "torchrun.supervisor.launchers"
+REGISTRY_POLICES_METADATA_GROUP_KEY = "torchrun.supervisor.policies"
+
+logger = get_logger(__name__)
+
+
+from dataclasses import dataclass
+
+
+@dataclass
+class ProcessConfig:
+    """Process configuration
+
+    Args:
+        run_path: path to the script to run
+        no_python: if True, run the script directly without python
+        module: if True, run the script as a python module
+    """
+
+    run_path: Optional[str]
+    no_python: Optional[bool]
+    module: Optional[bool]
+
+
+@dataclass
+class TimeoutConfig:
+    """Timeout configuration
+
+    Args:
+        join_timeout: timeout for hostmanagers to join
+        close_timeout: timeout for graceful process exit
+    """
+
+    join_timeout: float = 300
+    close_timeout: float = 60
+
+
+@dataclass
+class JobConfig:
+    """Job configuration
+
+    Args:
+        min_nodes: minimum number of nodes to allocate
+        max_nodes: maximum number of nodes to allocate
+        nproc_per_node: number of processes per node
+        max_restarts: maximum number of restarts
+        timeouts: timeout configuration
+        training_script: path to the training script
+        training_script_args: arguments to the training script
+
+    """
+
+    min_nodes: int
+    max_nodes: int
+    nproc_per_node: int
+    max_restarts: int
+    timeouts: TimeoutConfig
+    training_script: str
+    training_script_args: List[str]
+
+
+@dataclass
+class SupervisorConfig:
+    """Supervisor configuration
+
+    Args:
+        proc: Process configuration/behavior for process launching on the hosts
+        job: Job configuration
+        run_id: run id
+        port: port that supervisor listens on
+        host: host that supervisor runs on
+        root: if True, run as supervision root (not a hostmanager)
+
+    """
+
+    proc: ProcessConfig
+    job: JobConfig
+
+    run_id: Optional[str] = None
+    port: Optional[int] = None
+    host: Optional[str] = None
+    root: Optional[bool] = None
+
+
+Policy = Callable[[Context, SupervisorConfig], None]
+Launcher = Callable[[Policy, SupervisorConfig], None]
+
+
+class ProcessFailure(Exception):
+    def __init__(self, blamed: Process):
+        super().__init__()
+        self.blamed = blamed
+
+
+def _allocate_hosts(ctx: Context, conf: SupervisorConfig):
+    hosts = ctx.request_hosts(conf.job.max_nodes)
+    hosts_with_hostname: Sequence[Tuple[Host, Future[str]]] = wait_on(
+        hosts, lambda host: host.hostname(), conf.job.timeouts.join_timeout
+    )
+    connected_hosts = [host for host, _ in hosts_with_hostname]
+    return connected_hosts
+
+
+# TODO: extract Context interface as a public API
+def _torchrun_policy(ctx: Context, conf: SupervisorConfig):
```

Comment:
since this will likely grow to support more of torchrun options maybe we should put it in its own file `policy.py` or something.